feat: add main unittest and support retry, paged query, adding tags

ritch2022 opened the pull request:
changes:
1. add unittest for skeleton: provider, client and client, the coverage as belows: 
<img width="884" alt="image" src="https://user-images.githubusercontent.com/112048684/187900936-2f9eac12-3c85-4c64-a7a0-6ccbbe582b76.png">

3. some new functionalities:
* support retry
* support paged query
* support adding tag

ritch2022 commented:
not sure why CI / Tests failed
something wrong with  __pycache__ or *.pyc files?

kapilt commented:
its more to do with the min python version being specified in the Tencent provider I believe, it needs to go down to 3.7, per upstream sdk that should be fine, its currently specifying min 3.9

ritch2022 commented:
some info:
1. changed the Makefile and re-ordered the tencentcloud and azure, otherwise workflow "CI / Docs" will fail, error msg will be something like "no module typing-extensions"
2. changed the test files' name to avoid duplicated name

kapilt commented on tools/c7n_tencentcloud/c7n_tencentcloud/query.py:
```diff
@@ -2,106 +2,270 @@
+                })
+        return result
+
+    def get_resource6(self, resource_id):
```
interesting, looks like arn by any other name. is there a another name for it, perhaps we call it a resource_descriptor  or `qcs`

MaDongdong99 replied:
Yes, it is named qcs for tencentcloud as arn. Will rename it.

kapilt commented:
it looks like the test infrastructure here is being setup to mock the results, we generally eschew that by and large as its low fidelity to actual behavior or responses over time, in favor of flight recording test data, where in the actual responses are recorded as test data on disk, and replayed for test execution. it looks like the sdk is based on requests, and the extant library we utilize for that in c7n_azure is vcrpy and vcr_unittest https://vcrpy.readthedocs.io/en/latest/

kapilt commented on tools/c7n_tencentcloud/tests/conftest.py:
```diff
@@ -0,0 +1,30 @@
+
+@pytest.fixture
+def mock_env_aksk(monkeypatch):
+    monkeypatch.setenv("TENCENTCLOUD_SECRET_ID", "xxx")
```
for the test environment variables on creds we generally use the test.env file in the root dir. that allows easier switch between actual and fake credentials when exercising tests.

kapilt commented:
fwiw, this does bring coverage up to about 90% as is

```
❯ pytest --cov c7n_tencentcloud tests
Test session starts (platform: darwin, Python 3.10.0, pytest 7.1.2, pytest-sugar 0.9.5)
rootdir: /Users/kapilt/projects/cloud-custodian, configfile: tox.ini
plugins: xdist-2.5.0, forked-1.4.0, sugar-0.9.5, terraform-0.6.1, cov-3.0.0
collecting ... 
 tools/c7n_tencentcloud/tests/test_provider.py ✓✓✓✓                                                                                                                                                     15% █▋        
 tools/c7n_tencentcloud/tests/test_tc_client.py ✓✓✓✓✓✓✓✓✓✓✓✓                                                                                                                                            62% ██████▎   
 tools/c7n_tencentcloud/tests/test_tc_query.py ✓✓✓✓✓✓✓✓✓✓                                                                                                                                              100% ██████████

---------- coverage: platform darwin, python 3.10.0-final-0 ----------
Name                                         Stmts   Miss  Cover
----------------------------------------------------------------
c7n_tencentcloud/__init__.py                     0      0   100%
c7n_tencentcloud/actions/__init__.py             0      0   100%
c7n_tencentcloud/actions/tags.py                 4      0   100%
c7n_tencentcloud/client.py                      66      1    98%
c7n_tencentcloud/entry.py                        5      5     0%
c7n_tencentcloud/filters/__init__.py             0      0   100%
c7n_tencentcloud/output.py                       0      0   100%
c7n_tencentcloud/provider.py                    24      1    96%
c7n_tencentcloud/query.py                      141     11    92%
c7n_tencentcloud/resources/__init__.py           0      0   100%
c7n_tencentcloud/resources/cvm.py               11     11     0%
c7n_tencentcloud/resources/resource_map.py       1      0   100%
c7n_tencentcloud/resources/tag.py               19      1    95%
c7n_tencentcloud/utils.py                        5      0   100%
----------------------------------------------------------------
TOTAL                                          276     30    89%


Results (10.42s):
      26 passed
```

I think we probably could use one more fixture here re bringing in the c7n.testing fixture, but that can be in a followup.

kapilt reviewed (approved):
lgtm
